ports/servoshell/egl/app_state.rs

```diff
@@ -332,6 +333,26 @@ impl WebViewDelegate for RunningAppState {
     }
 }
 
+#[derive(Default)]
+pub struct VsyncRefreshDriver {
+    start_frame_callback: RefCell<Option<Box<dyn Fn() + Send>>>,
+}
+
+impl VsyncRefreshDriver {
+    fn notify_vsync(&self) {
+        let Some(start_frame_callback) = self.start_frame_callback.borrow_mut().take() else {
+            return;
+        };
+        start_frame_callback();
+    }
+}
+
+impl RefreshDriver for VsyncRefreshDriver {
+    fn observe_next_frame(&self, new_start_frame_callback: Box<dyn Fn() + Send>) {
+        *self.start_frame_callback.borrow_mut() = Some(new_start_frame_callback);
```

Comment:
Should we log a warining if `start_frame_callback` is `Some` already? Does it indicate some sort of 'dropped' frame that we should know about?

Reply:
I can add a warning here. It should theoretically never happen, but I suppose the embedder should distrust the Servo API and handle this case gracefully.